crates/ty_python_semantic/resources/mdtest/literal/collections/generator_expressions.md

```diff
@@ -1,6 +1,63 @@
 # Generator expressions
 
+## Basic
+
+We infer specialized `GeneratorType` instance types for generator expressions:
+
+```py
+# revealed: GeneratorType[int, None, None]
+reveal_type(x for x in range(10))
+
+# revealed: GeneratorType[tuple[int, str], None, None]
+reveal_type((x, str(y)) for x in range(3) for y in range(3))
+```
+
+When used in a loop, the yielded type can be inferred:
+
 ```py
-# revealed: GeneratorType[@Todo(generator expression yield type), @Todo(generator expression send type), @Todo(generator expression return type)]
-reveal_type((x for x in range(42)))
+squares = (x**2 for x in range(10))
+
+for s in squares:
+    reveal_type(s)  # revealed: int
+```
+
+When explicitly annotated, the yield type can be widened:
+
+```py
+from types import GeneratorType
+
+maybe_ints: GeneratorType[int | None, None, None] = (x**2 for x in range(10))
+
+def _():
+    reveal_type(maybe_ints)  # revealed: GeneratorType[int | None, None, None]
+
+    for s in maybe_ints:
+        reveal_type(s)  # revealed: int | None
+```
```

Comment:
I don't think this demonstrates anything specific about generator expressions, does it? I think this is just our general behaviour that we always respect the declared type of a variable over its narrowed local type when we observe the variable from an external scope (in this case, we're viewing the global variable from the perspective of the function scope).

If I add a `reveal_type` below the function scope, in the global scope, it looks like we reveal the narrowed type of `GeneratorType[int, None, None]` (which I think is correct, since `GeneratorType` is covariant -- there's no benefit from avoiding narrowing the type here)